private/buf/buflsp/buflsp.go

```diff
@@ -0,0 +1,634 @@
+	"go.lsp.dev/jsonrpc2"
+	"go.lsp.dev/protocol"
+	"go.uber.org/zap"
+)
+
+var (
+	// wellKnownTypesDescriptorProtoPath is the path of the descriptor.proto well-known types file.
+	wellKnownTypesDescriptorProtoPath = normalpath.Join("google", "protobuf", "descriptor.proto")
+	// lspWellKnownTypesCacheRelDirPath is the relative path to the cache directory for materialized
+	// well-known types .proto files.
+	lspWellKnownTypesCacheRelDirPath = normalpath.Join("lsp", "wkt")
+	// v3CacheModuleRelDirPath is the relative path to the cache directory in its newest iteration.
+	// NOTE: This needs to be kept in sync with module_data_provider.
+	v3CacheModuleRelDirPath = normalpath.Join("v3", "module")
+)
+
+// NewServer returns a new LSP server for the jsonrpc connection.
+func NewServer(
+	ctx context.Context,
+	jsonrpc2Conn jsonrpc2.Conn,
+	container appext.Container,
+	controller bufctl.Controller,
+) (protocol.Server, error) {
+	return newServer(ctx, jsonrpc2Conn, container, controller)
+}
+
+// *** PRIVATE ***
+
+type server struct {
+	nopServer
+
+	jsonrpc2Conn            jsonrpc2.Conn
+	logger                  *zap.Logger
+	container               appext.Container
+	controller              bufctl.Controller
+	lintHandler             buflint.Handler
+	breakingHandler         bufbreaking.Handler
+	fileCache               map[string]*fileEntry
+	fileWatcher             *fsnotify.Watcher
+	wellKnownTypesModuleSet bufmodule.ModuleSet
+	wellKnownTypesBucket    bufmodule.ModuleReadBucket
+
+	folders   []protocol.WorkspaceFolder
+	clientCap protocol.ClientCapabilities
+
+	lock sync.Mutex
+}
+
+func newServer(
+	ctx context.Context,
+	jsonrpc2Conn jsonrpc2.Conn,
+	container appext.Container,
+	controller bufctl.Controller,
+) (*server, error) {
+	logger := container.Logger()
+	tracer := tracing.NewTracer(container.Tracer())
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	wellKnownTypesModuleSet, err := wellKnownTypesModuleSet(ctx, logger)
+	if err != nil {
+		return nil, err
+	}
+	wellKnownTypesBucket := bufmodule.ModuleSetToModuleReadBucketWithOnlyProtoFiles(
+		wellKnownTypesModuleSet,
+	)
+	buflsp := &server{
+		jsonrpc2Conn:            jsonrpc2Conn,
+		logger:                  logger,
+		container:               container,
+		controller:              controller,
+		lintHandler:             buflint.NewHandler(logger, tracer),
+		breakingHandler:         bufbreaking.NewHandler(logger, tracer),
+		fileCache:               make(map[string]*fileEntry),
+		fileWatcher:             watcher,
+		wellKnownTypesModuleSet: wellKnownTypesModuleSet,
+		wellKnownTypesBucket:    wellKnownTypesBucket,
+	}
+	go func() {
+		for event := range buflsp.fileWatcher.Events {
+			if event.Op&fsnotify.Write == fsnotify.Write {
+				buflsp.lock.Lock()
+				if entry, ok := buflsp.fileCache[event.Name]; ok {
+					if entry.moduleSet != nil {
+						if err := buflsp.refreshImage(context.Background(), entry.moduleSet, entry.bucket); err != nil {
```

Comment:
Why are we using `context.Background()` and not the input `ctx` here?

Reply:
I am not 100% sure, seems like the reason is because it runs on a separate goroutine, but I think in this case that's perfectly fine since the goroutine is not meant to outlive the server that the context is for. I switched it to use `ctx`.